### DISCUSSION TITLE
JavaScript(ed) Rules

### DISCUSSION THREAD

**Discussion by justinstroud-campmeeker** [pmd/pmd#5215] [Ideas]
Back again, and with an update:

After some discussion and consideration, I've decided to move forward with building out a library that allows developers to author rules in JavaScript. A couple of points:
- This will be another dependency that is separate (currently) from PMD itself. It allows you to include the library alongside PMD if you want to use it. This protects pmd-core from the added JS engine dependency.
- It will be language independent. Currently I've been testing the engine out with Apex, and I've had some great and surprising results. Nashorn is actually a blast to work with because of the tight JVM integration. It just knows what node is what without wrappers around PMD objects of any kind.
* The biggest challenge I have is how the library is structured. If it is an abstract class, then there are some challenges around inheritance. Let's take Apex for example:
   * How would a rule inherit from this class? It won't be the rule itself, because that already inherits from AbstractApexRule. Do I put it one level below, or create another class called AbstractApexJavaScriptRule? That's a tough question, as I'd need to both duplicate the AbstractApexRule's implementations and create a class that is not language agnostic. I could create a class that gets called from a rule. That solves that problem, but leaves others...like, how to implement something that covers all languages and allows you to use some of the features like targeting specific nodes. RootNode is great, but it's not exactly the most performant choice.

Looking for feedback here. I've done a PoC and the idea works. It's just the exact implementation that needs some help. 

I do believe that this would be a powerful addition to PMD, as it opens up rule authoring to a wider audience. The way the PoC works, for example, is very similar to the way the XPath rules work: it's all configuration. You just specify the class that calls the Script that is in the XML configuration, and the JavaScript does everything else: find violations and flag them. 

Thanks!

**Comment by oowekyala** [comment]
Regarding your point about the rule base class, why would we need to make a language-specific base class? XPathRule is language-independent, as you can specify the behavior of the rule with its properties. I expect the Javascript rule to have the same design, as the behavior of the rule can be configured by the script, just like the xpath expression. I think it would be a mistake to add new base classes for rules and we should avoid that unless we have no other option. 

> like, how to implement something that covers all languages and allows you to use some of the features like targeting specific nodes. 

Maybe as a first step just add a property that specifies which nodes the rule is supposed to visit, and use that when overriding `AbstractRule::buildTargetSelector`. This property could contain the xpath names of the nodes, as we already build an index of the AST with these names as a key.

**Reply by justinstroud-campmeeker** [reply]
> ...why would we need to make a language-specific base class? XPathRule is language-independent, as you can specify the behavior of the rule with its properties

That was sort of where I was going with it, sorry if I was unclear or otherwise. I only use Apex for an example because that's what I currently write Java rules for...it's where I saw an opportunity to open the rules up to a wider audience. Some of it is also just me thinking out loud. Spitballing, if you will.

I don't want to make anything language specific if possible. The XPathRule is a good starting point, I agree. That was actually my first thought, I'm just working out how that would be structured. Still in the small stages (time and all that permitting.)

**Reply by adangel** [reply]
In general, I think such a feature would be a nice addition for PMD. Thanks for creating the PoC and testing it!

About the dependency and separate module or directly in pmd-core: if we ship this with pmd-dist, then it basically doesn't make a difference - the additional dependency will be there, not matter what. Having this feature as a separate optional module might result in that it is not used, as it is too cumbersome to get it working (if adding new jar files into pmd-dist would be easy, then we wouldn't need a feature, where you don't need to create an additional jar file to implement a custom rule).

About nashorn and rhino and "JavaScript" in general: For parsing javascript, we use [Rhino](https://github.com/mozilla/rhino) in our pmd-javascript module. Looking at the last commits on [nashorn](https://github.com/openjdk/nashorn) vs. rhino, then I would say, that nashorn is not supported anymore, while rhino still is developed. In that sense, I would choose rhino over nashorn (both do the same - and [Rhinoceros](https://en.wikipedia.org/wiki/Rhinoceros) and Nashorn are the same species - Nashorn is the German translation for Rhino).
Regardless of whether we choose nashorn or rhino in the end: both suffer from the same problem - they only fully support EcmaScript 5.1 and the newer JavaScript features are not really working. I'm afraid, if we add the possibility to write rules in JavaScript, then we immediately get a bug report, that e.g. the spread operator doesn't work. We will need to make it dead clear, that we don't support modern JavaScript, but only 13 year old JavaScript (5.1 is from 2011 according to https://en.wikipedia.org/wiki/ECMAScript_version_history ).
One side effect of choosing Rhino over Nashorn is, that we already use Rhino - so we wouldn't add an additional dependency (to be exact, we would add it to pmd-core, which means, it's used by everybody. But e.g. maven-pmd-plugin already depends on pmd-javascript, so already fetches rhino anyway).

About the rule definition: looking at your Poc (here https://gist.github.com/justinstroudbah/17790f569c731d988ec62b238b93d7df ), we basically need to define, what the JavaScript snippet will do when evaluated. There are a couple of options:
* it will define a method "hasViolation" that will be invoked by PMD. It returns a boolean. PMD can only flag the root node.
* it will define a method "findViolatedNodes", that will be invoked by PMD. It returns a list of nodes, for which violations should be reported:
```javascript
function findViolatedNodes(rootNode) {
  let result = [];
  let nodes = rootNode.descendants().filter(function(n) { return n.getXPathNodeName() === 'MethodDeclaration'; }).toList();
  for (node in nodes) {
    result.push(node);
  }
  return result;
}
```
* it could return a object with a specific structure. Looking at ESLint rules here (https://eslint.org/docs/latest/extend/custom-rules ). We could return e.g.
```javascript
{
  "VariableDeclaration": function(node, rulecontext) {
    if (node.getName() === "foo") {
        rulecontext.addViolation(node);
    }
  },
  "MethodDeclaration": function(node, rulecontext) {
    // ...
  }
}
```
This would be kind of a visitor, that is returned. Might be a bit tricky to map to the concrete visitor class of a specific language without writing such a mapper for each language separately. Maybe it is first easier to implement the "rule chain" behavior first - which means, in this example, we would find all "ASTVariableDeclaration" nodes and call this function, then we would find all "ASTMethodDeclaration" nodes and call the other function.
Note: I passed in the rulecontext, so that javascript can directly report violations.
* We could define in Java an interface, that needs to be implemented by the javascript snippet. E.g. we could say, we have:
```java
public interface JavaScriptRule {
  public void apply(Node node, RuleContext ruleContext);
}
```
And in Javascript (nashorn):
```javascript
return new net.sourceforge.pmd.core.lang.rule.JavaScriptRule() {
  apply: function(node, ruleContext) {
    //...
  }
};
```

Nashorn - [Implementing Java Interfaces](https://docs.oracle.com/en/java/javase/14/scripting/using-java-scripts.html#GUID-CB28D34D-BEDD-49B6-ACC8-E1CCD82694F6)
It seems, that Rhino works the same [Implementing interfaces](http://web.archive.org/web/20160514071428/https://developer.mozilla.org/en-US/docs/Mozilla/Projects/Rhino/Embedding_tutorial#implementingInterfaces)


In general I think, we should provide only the simplest way - e.g. passing in the root node and rulecontext. And add support for visitors/rule chain later. I guess, even with this JavaScriptRule-interface, one can already do many things - as you showed, even the NodeStream methods can simply be called.

In order to have this integrated into the main pmd (or pmd-core) code base, we need
- unit tests
- documentation - this adds a 3rd possibility to write rules. https://docs.pmd-code.org/latest/pmd_userdocs_extending_writing_rules_intro.html#writing-new-rules only talks about two ways...
- think about, whether the RuleDesigner can be of any help/support writing javascript based rules
